components/logins/src/encryption.rs

```diff
@@ -160,6 +168,133 @@ impl KeyManager for StaticKeyManager {
+///         Ok("secret".to_string())
+///     }
+///
+///     fn on_authentication_success(&self) {
+///         println!("success");
+///     }
+///
+///     fn on_authentication_failure(&self) {
+///         println!("this did not work, please try again:");
+///     }
+/// }
+/// let key_manager = NSSKeyManager::new(Arc::new(MyPrimaryPasswordAuthenticator {}));
+/// assert_eq!(key_manager.get_key().unwrap().len(), 63);
+/// ```
+#[cfg(feature = "keydb")]
+#[derive(uniffi::Object)]
+pub struct NSSKeyManager {
+    primary_password_authenticator: Arc<dyn PrimaryPasswordAuthenticator>,
+}
+
+#[cfg(feature = "keydb")]
+impl NSSKeyManager {
+    /// Initialize new `NSSKeyManager` with a given `PrimaryPasswordAuthenticator`.
+    /// There must be a previous initializiation of NSS before initializing
+    /// `NSSKeyManager`, otherwise this panics.
+    pub fn new(primary_password_authenticator: Arc<dyn PrimaryPasswordAuthenticator>) -> Self {
+        assert_nss_initialized();
+        Self {
+            primary_password_authenticator,
+        }
+    }
+}
+
+/// Identifier for the logins key, under which the key is stored in NSS.
+#[cfg(feature = "keydb")]
+static KEY_NAME: &str = "as-logins-key";
+
+#[cfg(feature = "keydb")]
+impl KeyManager for NSSKeyManager {
+    fn get_key(&self) -> ApiResult<Vec<u8>> {
+        if authentication_with_primary_password_is_needed().map_err(|e: nss::Error| {
+            LoginsApiError::NSSAuthenticationError {
+                reason: e.to_string(),
+            }
+        })? {
+            let primary_password = self
+                .primary_password_authenticator
+                .get_primary_password()
+                .map_err(|e| LoginsApiError::AuthenticationError {
+                    reason: e.to_string(),
+                })?;
+
+            let mut result = authenticate_with_primary_password(&primary_password).map_err(
+                |e: nss::Error| LoginsApiError::NSSAuthenticationError {
+                    reason: e.to_string(),
+                },
+            )?;
+
+            if result {
+                self.primary_password_authenticator
+                    .on_authentication_success();
+            } else {
+                while !result {
+                    self.primary_password_authenticator
+                        .on_authentication_failure();
+                    let primary_password = self
+                        .primary_password_authenticator
+                        .get_primary_password()
+                        .map_err(|e| LoginsApiError::AuthenticationError {
+                            reason: e.to_string(),
+                        })?;
+
+                    result = authenticate_with_primary_password(&primary_password).map_err(
+                        |e: nss::Error| LoginsApiError::NSSAuthenticationError {
+                            reason: e.to_string(),
+                        },
```

Comment:
How do you feel about adding functions that wrap these NSS methods and map the error?  I think that could make the logic of this function more clear, but it's not a big deal either way.